renderer/BloomStage.cpp

```diff
@@ -0,0 +1,321 @@
+							downsampleShader( nullptr ), downsampleWithBrightPassShader( nullptr ), upsampleShader( nullptr ) {
+}
+
+void BloomStage::Init() {
+	bloomDownSamplers = globalImages->ImageFromFunction("Bloom Downsampling", CreateBloomDownSamplingBuffer);
+	bloomDownSamplers->ActuallyLoadImage();
+	bloomUpSamplers = globalImages->ImageFromFunction("Bloom Upsampling", CreateBloomUpSamplingBuffer);
+	bloomUpSamplers->ActuallyLoadImage();
+
+	qglGenFramebuffers(NUM_FILTER_STEPS, downsampleFBOs);
+	for (int i = 0; i < NUM_FILTER_STEPS; ++i) {
+		qglBindFramebuffer(GL_FRAMEBUFFER, downsampleFBOs[i]);
+		qglFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, bloomDownSamplers->texnum, i);
+	}
+	qglGenFramebuffers(NUM_FILTER_STEPS, upsampleFBOs);
+	for (int i = 0; i < NUM_FILTER_STEPS; ++i) {
+		qglBindFramebuffer(GL_FRAMEBUFFER, upsampleFBOs[i]);
+		qglFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, bloomUpSamplers->texnum, i);
+	}
+	qglBindFramebuffer(GL_FRAMEBUFFER, 0);
+
+	downsampleShader = programManager->Find("bloom_downsample");
+	if (downsampleShader == nullptr) {
+		downsampleShader = programManager->LoadFromGenerator("bloom_downsample", LoadBloomDownsampleShader);
+	}
+	downsampleWithBrightPassShader = programManager->Find("bloom_downsample_brightpass");
+	if (downsampleWithBrightPassShader == nullptr) {
+		downsampleWithBrightPassShader = programManager->LoadFromGenerator("bloom_downsample_brightpass", LoadBloomDownsampleWithBrightPassShader);
+	}
+	blurShader = programManager->Find("bloom_blur");
+	if (blurShader == nullptr) {
+		blurShader = programManager->LoadFromGenerator("bloom_blur", LoadBloomBlurShader);
+	}
+	upsampleShader = programManager->Find("bloom_upsample");
+	if (upsampleShader == nullptr) {
+		upsampleShader = programManager->LoadFromGenerator("bloom_upsample", LoadBloomUpsampleShader);
+	}
+}
+
+void BloomStage::Shutdown() {
+	if (downsampleFBOs[0] != 0) {
+		qglDeleteFramebuffers(NUM_FILTER_STEPS, downsampleFBOs);
+		downsampleFBOs[0] = 0;
+	}
+	if (upsampleFBOs[0] != 0) {
+		qglDeleteFramebuffers(NUM_FILTER_STEPS, upsampleFBOs);
+		upsampleFBOs[0] = 0;
+	}
+	if (bloomDownSamplers != nullptr) {
+		bloomDownSamplers->PurgeImage();
+		bloomDownSamplers = nullptr;
+	}
+	if (bloomUpSamplers != nullptr) {
+		bloomUpSamplers->PurgeImage();
+		bloomUpSamplers = nullptr;
+	}
+}
+
+extern GLuint fboPrimary;
+extern bool primaryOn;
+
+void BloomStage::ComputeBloomFromRenderImage() {
+	GL_PROFILE("BloomStage");
+
+	int expectedWidth = r_fboResolution.GetFloat() * glConfig.vidWidth / 2;
+	int expectedHeight = r_fboResolution.GetFloat() * glConfig.vidHeight / 2;
+
+	if (downsampleFBOs[0] != 0 && (bloomDownSamplers->uploadWidth != expectedWidth || bloomDownSamplers->uploadHeight != expectedHeight)) {
+		// resolution changed, need to recreate our resources
+		Shutdown();
+	}
+
+	if (downsampleFBOs[0] == 0) {
+		Init();
+	}
+
+	qglClearColor(0, 0, 0, 0);
+	Downsample();
+	Blur();
+	Blur();
```

Comment:
I wonder what is done by Blur filter. I seems missing from original presentation. And why is it called twice?
Does it play nicely for upsampling later?

Reply:
I originally started with an approach that did blurring just by down- and upsampling, but I felt that the blur achieved by doing so was not strong enough. Blur performs an actual Gaussian Blur at the lowest downsampling level to get that smoother, wider blur. And basically it's done twice to increase the radius of the blur. I'll add some comments to explain the approach.